### PR TITLE
builtins: add repaired_descriptor builtin

### PR DESCRIPTION
**Pull request by postamar**
This commit adds a new crdb_internal.repaired_descriptor builtin
which makes use of the recently-added StripDanglingBackReferences method
on catalog.MutableDescriptor.

In conjunction with crdb_internal.unsafe_upsert_descriptor, this builtin
makes it possible to repair descriptors corrupted with dangling
back-references (back-references to descriptors or jobs which are
invalid) without needing to craft a descriptor surgery query for each
descriptor. It is therefore possible to repair this whole class of
corruptions in bulk.

This class of descriptor corruptions is particularly insidious in that
it is only caught by validation checks performed when writing new
versions of the descriptor (i.e. schema changes). This is deliberate, in
that we don't want this kind of corruption to interfere with DML queries
because those don't interact with back-references at all. On the other
hand, it defers detection of the corruption until a schema change or
a cluster upgrade is attempted.

Informs https://github.com/cockroachdb/cockroach/issues/104425.

Release note: None

### CODE REVIEW

**Comment by postamar** [comment]
The third commit is an opportunistic interface cleanup which isn't strongly related to the other two.

File: pkg/sql/sem/builtins/builtins.go [hunk 1]
```diff
@@ -5123,7 +5123,77 @@ DO NOT USE -- USE 'CREATE TENANT' INSTEAD`,
...
-				ret, err := evalCtx.CatalogBuiltins.DescriptorWithPostDeserializationChanges(ctx, []byte(s))
+				descIDAlwaysValid := func(id descpb.ID) bool {
+					return true
+				}
+				jobIDAlwaysValid := func(id jobspb.JobID) bool {
+					return true
+				}
+				ret, err := evalCtx.CatalogBuiltins.RepairedDescriptor(
+					ctx, []byte(s), descIDAlwaysValid, jobIDAlwaysValid,
+				)
```
Comment (postamar) [inline pkg/sql/sem/builtins/builtins.go:5134]: This change is a refactor.

**Review by fqazi — APPROVED** [review]
:lgtm: Nice work! Only minotr comments

> Reviewed 16 of 17 files at r1, 11 of 11 files at r4, 6 of 6 files at r5, 5 of 5 files at r6, 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105792#-:-NZ6gfnY6QrKNSmlOS6_:b-97haon)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @postamar)

___
*[`pkg/sql/catalog/tabledesc/table_desc_builder.go` line 241 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105792#-NZ6cxC88psgQR6MqBLK:-NZ6cxC88psgQR6MqBLL:b-m96c9x) ([raw file](https://github.com/cockroachdb/cockroach/blob/edde46042c3bec8470427cc288ae1ef9a11bdcc3/pkg/sql/catalog/tabledesc/table_desc_builder.go#L241)):*
> ```Go
> 		}
> 		if sliceIdx < len(tdb.maybeModified.MutationJobs) {
> 			tdb.maybeModified.MutationJobs = tdb.maybeModified.MutationJobs[:sliceIdx]
> ```

Sometimes we have leftover mutations in these cases, but this gives us the infrastructure to deal with those, too, in simple cases. Not saying this is something for this PR.
___
*[`pkg/sql/catalog/dbdesc/database_test.go` line 420 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105792#-NZ6eLNi9skggfHwJu8C:-NZ6eLNi9skggfHwJu8D:b-h1zuqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/edde46042c3bec8470427cc288ae1ef9a11bdcc3/pkg/sql/catalog/dbdesc/database_test.go#L420)):*
> ```Go
> 
> 	for i, test := range testData {
> 		t.Run(fmt.Sprintf("#%02d", i+1), func(t *testing.T) {
> ```

Maybe just give them names to make it easier to diagnose in the future.

**Comment by chengxiong-ruan** [comment]
nice work, I think it'd be useful if we backport it to 23.1 :)

**Comment by postamar** [comment]
Thanks for the reviews.

> Sometimes we have leftover mutations in these cases, but this gives us the infrastructure to deal with those, too, in simple cases. Not saying this is something for this PR.

Very good point! I added this in this PR and added a test case to cover it. This has annoyed us in the past.

> Maybe just give them names to make it easier to diagnose in the future.

Done.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105792)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[`-- commits` line 55 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/105792#-NZ7dW551wVb0eNlMm0p):*
hm, but haven't all the system descriptors out in the wild already had their mod_time adjusted thanks to `runPostDeserializationChangesOnAllDescriptors`?
___
*[`pkg/sql/sem/builtins/builtins.go` line 5134 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105792#gh-1246526325:-NZ7cfbI7BD1cpKBi8nn:b-bas9lj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e21355c976e762c1468910a454ea48adc0664e3e/pkg/sql/sem/builtins/builtins.go#L5134)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This change is a refactor.
</blockquote></details>

is it? i could be dense here, but where does `RunPostDeserialiationChanges` get called now? (neither `CatalogBuiltins.RepairedDescriptor` nor `StripDanglingBackReferences` seems to do that)

**Comment by postamar** [comment]
> hm, but haven't all the system descriptors out in the wild already had their mod_time adjusted thanks to runPostDeserializationChangesOnAllDescriptors?

When you bootstrap and never upgrade cluster versions, the system database descriptors don't have any of their timing fields set (mod_time, create_as_of_time, etc.) because when bootstrapping we just `Put` some predefined bytes blob with those fields unset.

> is it? i could be dense here, but where does RunPostDeserialiationChanges get called now? 

It gets called in `descbuilder.FromBytesAndMVCCTimestamp`. It's not super obvious.

**Comment by rafiss** [comment]
> When you bootstrap and never upgrade cluster versions, the system database descriptors don't have any of their timing fields set (mod_time, create_as_of_time, etc.) because when bootstrapping we just Put some predefined bytes blob with those fields unset.

Right, but aren't there clusters out in the wild where the system descriptors were bootstrapped in 22.1, then during the upgrade to 22.2 upgrade, `runPostDeserializationChangesOnAllDescriptors` was executed, which would set the mod_time on all the system descriptors?

**Comment by postamar** [comment]
I think so, yes. At least, that's the case for the create_as_of_time field. In any case, it's benign, I think.

**Pull request merged**